src/System.Linq/src/System/Linq/Take.SpeedOpt.cs

```diff
@@ -12,5 +13,41 @@ public static partial class Enumerable
             source is IPartition<TSource> partition ? partition.Take(count) :
             source is IList<TSource> sourceList ? (IEnumerable<TSource>)new ListPartition<TSource>(sourceList, 0, count - 1) :
             new EnumerablePartition<TSource>(source, 0, count - 1);
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(count > 0);
+
+            if (source is IPartition<TSource> partition)
+            {
+                int length = partition.GetCount(true);
+                if (length >= 0)
+                {
+                    if (length > count)
+                    {
+                        return partition.Skip(length - count);
+                    }
+                    else
+                    {
+                        return partition;
+                    }
+                }
```

Comment:
This could just be:
```C#
if (length > 0)
{
    return partition.Skip(length - count);
}
else if (length == 0)
{
    return partition;
}
```

Reply:
Right. Also added a checkup if Skip is needed/possible.